### PR TITLE
(MODULES-8603) Ignore .keep_* files

### PR DESCRIPTION
**Pull request by ekohl**
On Gentoo there's always a `.keep_<package>-<slot>` (e.g., `.keep_cronbase-0`) file inside the cron directory to ensure it's not removed with rmdir. Since usernames are very unlikely to start with `.keep_`, we can safely filter out these hidden files.

### CODE REVIEW

**Comment by binford2k** [comment]
While in practice, that's infinitely close to being accurate, `.` is actually valid as the first character of a POSIX username. I don't know how much it truly matters though, since `adduser` etc. won't do it.

http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap03.html#tag_03_435

> A string that is used to identify a user; see also User Database. To be portable across systems conforming to POSIX.1-2017, the value is composed of characters from the portable filename character set. The <hyphen-minus> character should not be used as the first character of a portable user name.

The valid characters are:

```
A B C D E F G H I J K L M N O P Q R S T U V W X Y Z
a b c d e f g h i j k l m n o p q r s t u v w x y z
0 1 2 3 4 5 6 7 8 9 . _ -
```

File: lib/puppet/provider/cron/crontab.rb [hunk 1]
```diff
@@ -283,7 +283,7 @@ def self.enumerate_crontabs
     return unless File.readable?(CRONTAB_DIR)
     Dir.foreach(CRONTAB_DIR) do |file|
       path = "#{CRONTAB_DIR}/#{file}"
-      yield(file) if File.file?(path) && File.writable?(path)
+      yield(file) if File.file?(path) && File.writable?(path) && !file.start_with?('.')
```
Comment (binford2k) [inline lib/puppet/provider/cron/crontab.rb]: In context of my other comment, should this check for `start_with?('.keep_')`?

**Comment by bastelfreak** [comment]
@binford2k oh lol, :+1: for being such an awesome nerd to even think about checking this :D And I was already proud about myself that I checked if `start_with()` is a valid method. I didn't even consider if `.` is valid.

**Comment by anthonyryan1** [comment]
Excluding `.keep_`  would certainly resolve the issues on Gentoo, I'm inclined to say that's the correct approach here given the . username prefix concerns. I'm not aware of any other distro that uses dotfiles in this directory.

**Comment by ekohl** [comment]
And this is why you write a commit message with the problem and your reasoning for the solution. Others can find a mistake in it and suggest a better solution :)

**Comment by joshcooper** [comment]
Thanks for the contribution @ekohl. The change seems fine, I assume no one would realistically have a file named `.keep_` and expect puppet to delete it?

To get this merged could you:
* Create a JIRA ticket in the `MODULES` project
* Update the subject line in the commit to reference the ticket, eg `(MODULES-NNN) Ignore ..`
* Add a unit test that keep files are indeed skipped

Thanks!

**Comment by melissa** [comment]
Is this something that should get in prior to the upcoming release @puppetlabs/platform-os ?

**Comment by branan** [comment]
@melissa I don't think so (given modules are cheap to release) - I think we can land this async either when the original author pops back up to address @joshcooper's requests, or when we make the time to clean up this commit ourselves. It doesn't need to end up in the agent vendored modules, since we don't ship Gentoo packages ourselves.

**Comment by ekohl** [comment]
Updated with a modules ticket.

**Comment by binford2k** [comment]
@joshcooper I've been poking at this and I can't come up with a way to test it that actually tests anything effectively :( Any suggestions?

**Comment by joshcooper** [comment]
I'd add something like the following (not sure it works) to https://github.com/puppetlabs/puppetlabs-cron_core/blob/1.0.0/spec/unit/provider/cron/parsed_spec.rb#L93-L109

```ruby
it 'ignores .keep_* files' do
  files = tabs + %w[.keep_cronbase-0]
  Dir.expects(:foreach).multiple_yields(*files)
  expect(described_class.targets).to eq(tabs)
end
```

**Comment by ekohl** [comment]
It's been a while, but I've gotten around to adding some tests.

**Comment by ekohl** [comment]
Ping?

**Comment by GabrielNagy** [comment]
Hi @ekohl, can you rebase on the current master to get https://github.com/puppetlabs/puppetlabs-cron_core/commit/b9944b0fd1b6e3e64af7599ec185eab49f0b1a0e into your tree so I can trigger an acceptance run of this change? Otherwise this should be good to merge. Thanks for the contribution!

**Comment by ekohl** [comment]
Rebased

**Pull request merged**